core/field_variable.js

```diff
@@ -135,37 +121,41 @@ Blockly.FieldVariable.prototype.setValue = function(newValue) {
  * @this {Blockly.FieldVariable}
  */
 Blockly.FieldVariable.dropdownCreate = function() {
-  var variableNameList = [];
-  if (this.sourceBlock_ && this.sourceBlock_.workspace) {
+  var variableModelList = [];
+  var name = this.getText();
+  // Don't create a new variable if there is nothing selected.
+  var createSelectedVariable = name ? true : false;
+  var workspace = null;
+  if (this.sourceBlock_) {
+    workspace = this.sourceBlock_.workspace;
+  }
+
+  if (workspace) {
     // Get a copy of the list, so that adding rename and new variable options
     // doesn't modify the workspace's list.
-
-    var variableModelList = this.sourceBlock_.workspace.getVariablesOfType('');
-    for (var i = 0; i < variableModelList.length; i++) {
-      variableNameList.push(variableModelList[i].name);
+    var variableModelList = workspace.getVariablesOfType('');
+    for (var i = 0; i < variableModelList.length; i++){
+      if (createSelectedVariable &&
+          goog.string.caseInsensitiveEquals(variableModelList[i].name, name)) {
+        createSelectedVariable = false;
+        break;
+      }
     }
   }
   // Ensure that the currently selected variable is an option.
-  var name = this.getText();
-  if (name && variableNameList.indexOf(name) == -1) {
-    variableNameList.push(name);
+  if (createSelectedVariable && workspace) {
+    var newVar = workspace.createVariable(name);
+    variableModelList.push(newVar);
   }
-  variableNameList.sort(goog.string.caseInsensitiveCompare);
-
-  this.renameVarItemIndex_ = variableNameList.length;
-  variableNameList.push(Blockly.Msg.RENAME_VARIABLE);
-
-  this.deleteVarItemIndex_ = variableNameList.length;
-  variableNameList.push(Blockly.Msg.DELETE_VARIABLE.replace('%1', name));
-  // Variables are not language-specific, use the name as both the user-facing
-  // text and the internal representation.
+  variableModelList.sort(Blockly.VariableModel.compareByName);
   var options = [];
-  for (var i = 0; i < variableNameList.length; i++) {
-    // TODO(marisaleung): Set options[i] to [name, uuid]. This requires
-    // changes where the variable gets set since the initialized value would be
-    // id.
-    options[i] = [variableNameList[i], variableNameList[i]];
+  for (var i = 0; i < variableModelList.length; i++) {
+    // Set the uuid as the internal representation of the variable.
+    options[i] = [variableModelList[i].name, variableModelList[i].getId()];
   }
+  options.push([Blockly.Msg.RENAME_VARIABLE, Blockly.RENAME_VARIABLE_ID]);
+  options.push([Blockly.Msg.DELETE_VARIABLE.replace('%1', name),
+               Blockly.DELETE_VARIABLE_ID]);
```

Comment:
eslint might complain about this spacing. Might need to add one more space to line 158.